linalg/ginkgo.hpp

```diff
@@ -107,169 +309,334 @@ struct ResidualLogger : gko::log::Logger
-    * ```
-    * auto ref = gko::create<gko::ReferenceExecutor>();
-    * ```
-    *
-    * The following code snippet demonstrates the using of the OpenMP executor
-    * to create a solver which would use the OpenMP paradigm to the solve the
-    * system on the CPU.
+    */
+   GinkgoPreconditioner(GinkgoExecutor &exec);
+
+   /**
+    * Destructor.
+    */
+   virtual ~GinkgoPreconditioner() = default;
+
+   /**
+    * Generate the preconditioner for the given matrix @p op.
+    * Calling this function is only required when creating a
+    * preconditioner for use with another MFEM solver; to use with
+    * a Ginkgo solver, get the LinOpFactory  pointer through @p GetFactory()
+    * and pass to the Ginkgo solver constructor.
+    */
+   virtual void SetOperator(const Operator &op);
+
+   /**
+    * Apply the preconditioner to input vector @p x, with out @p y.
+    */
+   virtual void Mult(const Vector &x, Vector &y) const;
+
+   /**
+    * Return a pointer to the LinOpFactory that will generate the preconditioner
+    * with the parameters set through the specific constructor.
+    */
+   const std::shared_ptr<gko::LinOpFactory> GetFactory() const
+   {
+      return this->precond_gen;
+   };
+
+   /**
+    * Return a pointer to the generated preconditioner for a specific matrix
+    * (that has previously been sith with @p SetOperator).
+    */
+   const std::shared_ptr<gko::LinOp> GetGeneratedPreconditioner() const
+   {
+      return this->generated_precond;
+   };
+
+   /**
+    * Return whether this GinkgoPreconditioner object has an explicitly-
+    * generated preconditioner, built for a specific matrix.
+    */
+   bool HasGeneratedPreconditioner() const
+   {
+      return this->has_generated_precond;
+   };
+
+protected:
+   /**
+    * The Ginkgo generated solver factory object.
+    */
+   std::shared_ptr<gko::LinOpFactory> precond_gen;
+
+   /**
+    * Generated Ginkgo preconditioner for a specific matrix, created through
+    * @p SetOperator(), or a wrapped MFEM preconditioner.
+    * Must exist to use @p Mult().
+    */
+   std::shared_ptr<gko::LinOp> generated_precond;
+
+   /**
+    * The execution paradigm in Ginkgo. The choices are between
+    * `gko::OmpExecutor`, `gko::CudaExecutor` and `gko::ReferenceExecutor`
+    * and more details can be found in Ginkgo's documentation.
+    */
+   std::shared_ptr<gko::Executor> executor;
+
+   /**
+    * Whether or not we have generated a specific preconditioner for
+    * a matrix.
+    */
+   bool has_generated_precond;
+
+};
+
+/**
+* This class forms the base class for all of Ginkgo's iterative solvers.  The
+* various derived classes only take the additional data that is specific to them
+* and solve the given linear system. The entire collection of solvers that
+* Ginkgo implements is available at the Ginkgo documentation and manual pages,
+* https://ginkgo-project.github.io/ginkgo/doc/develop.
+*
+* @ingroup GinkgoWrappers
+*/
+class GinkgoIterativeSolver : public Solver
```

Comment:
Could this inherit from `mfem::IterativeSolver`?

Reply:
Hmmm...it's not clear to me that it could, at the moment.  I see `SetPreconditioner` is virtual, so we could maybe make an implementation that would work properly (it would likely mean changing the current constructors, which take the preconditioner as a parameter--we probably don't want two ways to set the preconditioner?).  I'm more concerned about `SetMonitor`, `SetMaxIter`, etc., which aren't virtual, so we can't reimplement them.  They would look like they were available for the Ginkgo solvers, but of course calling them wouldn't actually do anything to change the solver itself.

Reply:
The values set by non-virtual methods or read from such methods post-solve (like max iterations, tolerances, recording convergence status, saving the last norm, warning that monitors are not supported, if set, etc) can be used/set directly in `Mult`. The general advantage of doing that is the unified iterative solver interface presented to users, at least for common options.

Reply:
I've been thinking a lot more about this.  I think it would be doable, but it would require a significant overhaul of the whole interface.  Due to the way the Ginkgo solvers are generated, I don't really see any way around that...though looking more into this has reminded me that I might want to make changes so that a generated solver is stored (rather than just its factory) in case users want to call `Mult` multiple times...which seems like a very reasonable use case.

I am also wondering about the `IterativeSolverMonitor` and whether that is something Ginkgo would have to be able to interface with?  That's the only part of the `IterativeSolver` class that I'm not sure about whether or not it could work with Ginkgo.

It seems most third-party libraries' solvers (e.g. SuperLU, PETSc, hypre, STRUMPACK) inherit from `Solver` rather than `IterativeSolver`; is there currently a plan to shift everything to `IterativeSolver` (or as much as can be shifted)?

Reply:
Hi @nbeams, supporting MFEM's iterative solver interface (at some level) could be nice but I don't think we want to require that. In my post above I was just pointing out that closer integration might be possible and nice to have in general. However, you should feel free to tweak, or leave as is, Ginkgo's integration into MFEM as you think is appropriate. In particular, I don't think you should worry at all about supporting things like `IterativeSolverMonitor` -- this is very much implementation dependent option (it was meant to be used in MFEM classes really) which you can ignore. For example PETSc classes have their own ways of "monitoring" things.

Regarding the other solver libraries: I have not heard of any plans to change their current integration approach.

Reply:
Oh, I see, you'd like to add functions with the same names (for getting and setting) as those used in `IterativeSolver` where possible, without actually inheriting from `IterativeSolver`? 
This might still require some tweaking but should be simpler to implement with Ginkgo.  I will think some more about this over the weekend.
I was also trying to keep the current interface as close as possible to what was already in MFEM (though it's still not exactly the same), but if that's not too important, then we could definitely change it so that some of those properties (tolerance, max iterations, preconditioner, etc.) are set through functions rather than through arguments to the constructor...

Reply:
> Oh, I see, you'd like to add functions with the same names (for getting and setting) as those used in `IterativeSolver` where possible, without actually inheriting from `IterativeSolver`?

No, I wasn't thinking of that. 😁 

In https://github.com/mfem/mfem/pull/2068#discussion_r618823235, I was thinking about the case where `GinkgoIterativeSolver` inherits from `mfem::IterativeSolver` as Will was suggesting.

Anyway, don't take my posts here as recommendations or requests for the PR (I haven't looked at the PR really) -- these were more like general thoughts. Sorry if I left the wrong impression.

Reply:
I think the easiest approach is to leave it as it is currently. Inheriting from `IterativeSolver` isn't critical. It would be nice to have if the implementation was easy, but since it sounds like the implementation would be quite involved it's probably not worth it.

Reply:
Well, since I was already thinking about it now, I've been playing around with an option in which all the Ginkgo solvers have their own `SetRelTol`, `SetAbsTol`, `SetMaxIter`, `SetPrintLevel`, `SetPreconditioner` (except for IR), `GetConverged`, `GetFinalNorm`, and `GetNumIterations` functions, while still inheriting from `Solver`.  The only difference here is that `SetPreconditioner` should take a `GinkgoPreconditioner` type specifically (but an MFEM preconditioner can be wrapped as a `GinkgoPreconditioner` and then passed in). 
The Ginkgo factory gets built at the `Mult` call, once we're sure all the parameters are set; the only downside is that I had to make some more member variables `mutable` in order for this to work (since `Mult` is marked as `const`, and I'm overriding it so I can't change that).  I also tried adding some logic to make sure the solver is only generated if something has changed since the last time it was used, but I haven't tested that yet.  Actually there are a few known todos left, but I think the core functionality is working. 

Anyway, since this would be such a major change, I wasn't sure if I should go ahead and push to this branch or if I should make a temporary branch first to show what I mean (in case we decide we don't like these changes and want to revert)? 

Another easy option is to forget the `Set` functions but go ahead and add the `Get` ones, which is trivial on its own.

Reply:
> Another easy option is to forget the `Set` functions but go ahead and add the `Get` ones, which is trivial on its own.

Perhaps we should go with this option to make everyone's lives easier. What you do think @YohannDudouit @v-dobrev?

Reply:
I don't have a strong preference - I'm okay with both options.

Reply:
I'm also wondering how important this use case is to MFEM/its users, in your opinion: being able to alter some of the things set by the `Set` functions after using the solver, and not just in the initial setup phase?  That is, if the Ginkgo interface is limited to its current setup (tolerance, max iterations, and preconditioner information is set at object creation and not changed -- only the Operator/matrix can be changed later with`SetOperator`), would that be particularly limiting to any known or likely use cases?  

During our meeting today, I was talking to the Ginkgo team at length about this issue.  They are currently having internal discussions that will relate to this, deciding what Ginkgo's policy should be about mutability of various solver parameters (like tolerances or preconditioners).  There is also a chance that the solvers will have a base class in the future which would make it simple to write a`GinkgoIterativeSolver` implementation of `Set*` functions to actually update the Ginkgo factory/generated solver, without having to push the factory update/regeneration into `Mult`.
That might not be possible until a later PR depending on how things go with Ginkgo, but it's something to keep in mind in terms of what the interface should be -- since I'm already changing the interface in this PR from the original Ginkgo wrappers, we'd obviously like to avoid changing it more in the future.

Reply:
Strictly speaking from a design point of view, these `Set` functions are unnecessary, as you correctly state the values set are not meant to change after being set once, the general consensus here is that [RAII](https://en.cppreference.com/w/cpp/language/raii) should be used whenever possible. The only valid reason to not use RAII is when it's hard to construct correctly initialized objects, which doesn't seem to be the case here.
On the other hand MFEM heavily uses this design pattern of creating invalid objects and then making them valid by using a bunch of `Set`, but I still think that we shouldn't add these `Set` methods, especially considering the trouble.

Reply:
There are use cases for adaptive solver tolerances, e.g. in some forms of Newton iteration: https://github.com/mfem/mfem/blob/5b894c6541c61b82e09641b08275d1a3ba6fdf89/linalg/solvers.cpp#L1773

Reply:
So, it seems it would be useful to add the `SetRelTol`, `SetAbsTol`, and `SetMaxIter` functions for Ginkgo, if possible.  Would it be okay to leave the preconditioner only in the constructor, or do you think users will want to be able to change that after the fact, as well?

I've been playing around with a couple newer ideas for adding the `Set` functions (without needing to push regeneration to `Mult`), and it could still depend a bit on what Ginkgo decides to do with respect to an open PR (https://github.com/ginkgo-project/ginkgo/pull/753)... we could make it work either way, it just might be simpler in one case than another.

In fc9a4bfdf66c0d45f08dd1216cda7e4305d59022 I went ahead and added the `Get` functions related to convergence.

Reply:
> Would it be okay to leave the preconditioner only in the constructor, or do you think users will want to be able to change that after the fact, as well?

Support for changing the preconditioner, after it has already been set, is not that useful IMO.

Reply:
Sounds good.  In 85666652aa4125405700fec4e9bb4721ecfdf4ee, I added `SetRelTol`, `SetAbsTol`, and `SetMaxIter` functions.  These parameters were also removed from the constructors of the Ginkgo solvers, so now the setup and usage in `ex1` looks like:

```
            Ginkgo::GinkgoExecutor exec(device);
            Ginkgo::IcPreconditioner ginkgo_precond(exec, "paric", 30);
            Ginkgo::CGSolver ginkgo_solver(exec, ginkgo_precond);
            ginkgo_solver.SetPrintLevel(print_lvl);
            ginkgo_solver.SetRelTol(1e-12);
            ginkgo_solver.SetAbsTol(0.0);
            ginkgo_solver.SetMaxIter(400);
            ginkgo_solver.SetOperator(*(A.Ptr()));
            ginkgo_solver.Mult(B, X);
```
            

We do not have `SetPreconditioner` for now, and the preconditioner is still passed to the constructor (if you are using one).  I think this departure from the "usual" MFEM interface is maybe a good thing, to highlight that the preconditioner must be a `GinkgoPreconditioner` (even if you are using an MFEM preconditioner, you must set the operator and create a wrapper for it before passing it to the Ginkgo solver, at the moment).  

I also have code ready to push to add `SetKDim` for the the GMRES solvers; I'm just waiting for Ginkgo PR https://github.com/ginkgo-project/ginkgo/pull/763 to be merged to develop, so the necessary function is available for the compressed basis GMRES as well as the standard one.

Reply:
The Ginkgo PR has been merged, so I added `SetKDim` in 1d8fd22fe042538e724d0e535b7131374ced9f8a .